File: innerlibs1/Printer/EscDaruma.vb
Imports System.Drawing
Imports System.Text
Imports InnerLibs.Printer
Imports InnerLibs.Printer.Command

Namespace EscDarumaCommands

    Public Class EscDaruma
        Implements IPrintCommand

#Region "Properties"

        Public ReadOnly Property ColsNomal As Integer Implements IPrintCommand.ColsNomal
            Get
                Return 48
            End Get
        End Property

        Public ReadOnly Property ColsCondensed As Integer Implements IPrintCommand.ColsCondensed
            Get
                Return 57
            End Get
        End Property

        Public ReadOnly Property ColsExpanded As Integer Implements IPrintCommand.ColsExpanded
            Get
                Return 25
            End Get
        End Property

        'Encoding.GetEncoding("IBM860")
        Public Property Encoding As Encoding Implements IPrintCommand.Encoding

#End Region



#Region "Methods"

        Public Function Separator() As Byte() Implements IPrintCommand.Separator
            Return Condensed(True).AddTextBytes(New String("-"c, ColsCondensed), Encoding).AddBytes(Condensed(False)).AddLF()
        End Function

        Public Function AutoTest() As Byte() Implements IPrintCommand.AutoTest
            Return New Byte() {28, "M"c.ToByte(), 254, 0}
        End Function

        Private Shared Function StoreQr(ByVal qrData As String, ByVal size As QrCodeSize) As IEnumerable(Of Byte)
            Dim length = qrData.Length + 3
            Dim b = CByte(length Mod 255)
            Dim b2 = CByte(length / 255)
            Return New Byte() {27, 106, 49}.AddBytes((New Byte() {27, 129})).AddBytes({b}).AddBytes({b2}).AddBytes({(size + 3).ToByte()}).AddBytes({"M"c.ToByte()})
        End Function

        Public Function Printqrdata(ByVal qrData As String) As Byte() Implements IPrintCommand.PrintQrData
            Return PrintQrData(qrData, QrCodeSize.Size0)
        End Function

        Public Function PrintQrData(ByVal qrData As String, ByVal qrCodeSize As QrCodeSize) As Byte() Implements IPrintCommand.PrintQrData
            Dim list = New List(Of Byte)()
            list.AddRange(StoreQr(qrData, qrCodeSize))
            list.AddRange(Encoding.GetBytes(qrData))
            Return list.ToArray()
        End Function

        Public Function FullCut() As Byte() Implements IPrintCommand.FullCut
            Return New Byte() {27, "m"c.ToByte()}
        End Function

        Public Function PartialCut() As Byte() Implements IPrintCommand.PartialCut
            Return New Byte() {27, "m"c.ToByte()}
        End Function

        Public Function Initialize() As Byte() Implements IPrintCommand.Initialize
            Return New Byte() {27, "@"c.ToByte()}
        End Function

        Public Function PrintImage(ByVal image As Drawing.Image, ByVal highDensity As Boolean) As Byte() Implements IPrintCommand.PrintImage
            Dim list = New List(Of Byte)()
            Dim bmp = New Bitmap(image)
            Dim send = "" & Microsoft.VisualBasic.ChrW(27) & Microsoft.VisualBasic.ChrW(51) & Microsoft.VisualBasic.ChrW(0)
            Dim data = New Byte(send.Length - 1) {}

            For i = 0 To send.Length - 1
                data(i) = Microsoft.VisualBasic.AscW(send(i))
            Next

            list.AddRange(data)
            data(0) = Microsoft.VisualBasic.AscW(Microsoft.VisualBasic.Strings.ChrW(0))
            data(1) = Microsoft.VisualBasic.AscW(Microsoft.VisualBasic.Strings.ChrW(0))
            data(2) = Microsoft.VisualBasic.AscW(Microsoft.VisualBasic.Strings.ChrW(0))
            Dim escBmp As Byte() = {&H1B, &H2A, &H0, &H0, &H0}
            escBmp(2) = Microsoft.VisualBasic.AscW("!"c)
            escBmp(3) = CByte(bmp.Width Mod 256)
            escBmp(4) = CByte(bmp.Width / 256)

            For i = 0 To bmp.Height / 24 + 1 - 1
                list.AddRange(escBmp)

                For j = 0 To bmp.Width - 1

                    For k = 0 To 24 - 1

                        If i * 24 + k < bmp.Height Then
                            Dim pixelColor = bmp.GetPixel(j, i * 24 + k)
                            If Not (pixelColor.R > 160 AndAlso pixelColor.G > 160 AndAlso pixelColor.B > 160) Then data(k / 8) += CByte(128 >> k Mod 8)
                            If highDensity Then Continue For
                            If pixelColor.R = 0 Then data(k / 8) += CByte(128 >> k Mod 8)
                        End If
                    Next

                    list.AddRange(data)
                    data(0) = Microsoft.VisualBasic.AscW(Microsoft.VisualBasic.Strings.ChrW(0))
                    data(1) = Microsoft.VisualBasic.AscW(Microsoft.VisualBasic.Strings.ChrW(0))
                    data(2) = Microsoft.VisualBasic.AscW(Microsoft.VisualBasic.Strings.ChrW(0))
                Next
            Next

            list.AddRange(New Byte() {27, "@"c.ToByte()})
            Return list.ToArray()
        End Function

        Public Function DoubleWidth2() As Byte() Implements IPrintCommand.DoubleWidth2
            Return New Byte() {27, 14, 0}
        End Function

        Public Function DoubleWidth3() As Byte() Implements IPrintCommand.DoubleWidth3
            Return New Byte() {27, 14, 0}
        End Function

        Public Function NormalWidth() As Byte() Implements IPrintCommand.NormalWidth
            Return New Byte() {20}
        End Function

        Public Function Bold(ByVal value As String) As Byte() Implements IPrintCommand.Bold
            Return Bold(True).AddTextBytes(value, Encoding).AddBytes(Bold(False)).AddLF()
        End Function

        Public Function Bold(ByVal state As Boolean) As Byte() Implements IPrintCommand.Bold
            Return If(state = True, New Byte() {27, "E"c.ToByte()}, New Byte() {27, "F"c.ToByte()})
        End Function

        Public Function Condensed(ByVal value As String) As Byte() Implements IPrintCommand.Condensed
            Return Condensed(True).AddTextBytes(value, Encoding).AddBytes(Condensed(False)).AddLF()
        End Function

        Public Function Condensed(ByVal state As Boolean) As Byte() Implements IPrintCommand.Condensed
            Return If(state = True, New Byte() {27, 15}, New Byte() {27, 18, 20})
        End Function

        Public Function Expanded(ByVal value As String) As Byte() Implements IPrintCommand.Expanded
            Return Expanded(True).AddTextBytes(value, Encoding).AddBytes(Expanded(False)).AddLF()
        End Function

        Public Function Expanded(ByVal state As Boolean) As Byte() Implements IPrintCommand.Expanded
            Return If(state = True, New Byte() {27, "w"c.ToByte(), 1}, New Byte() {27, "w"c.ToByte(), 0})
        End Function

        Public Function Italic(ByVal value As String) As Byte() Implements IPrintCommand.Italic
            Return Italic(True).AddTextBytes(value, Encoding).AddBytes(Italic(False)).AddLF()
        End Function

        Public Function Italic(ByVal state As Boolean) As Byte() Implements IPrintCommand.Italic
            Return If(state = True, New Byte() {27, "4"c.ToByte(), 1}, New Byte() {27, "4"c.ToByte(), 0})
        End Function

        Public Function Underline(ByVal value As String) As Byte() Implements IPrintCommand.Underline
            Return Underline(True).AddTextBytes(value, Encoding).AddBytes(Underline(False)).AddLF()
        End Function

        Public Function Underline(ByVal state As Boolean) As Byte() Implements IPrintCommand.Underline
            Return If(state = True, New Byte() {27, "-"c.ToByte(), 1}, New Byte() {27, "-"c.ToByte(), 0})
        End Function

        Public Function OpenDrawer() As Byte() Implements IPrintCommand.OpenDrawer
            Return New Byte() {27, "p"c.ToByte()}
        End Function

        Public Function Code128(ByVal code As String) As Byte() Implements IPrintCommand.Code128
            Return New Byte() {27, "b"c.ToByte, 5}.AddBytes((New Byte() {2})).AddBytes((New Byte() {50})).AddBytes((New Byte() {0})).AddTextBytes(code, Encoding).AddBytes(New Byte() {0}) ' Code Type
            ' Witdh
            ' Height
            ' If print code informed (1 print, 0 dont print)
        End Function

        Public Function Code39(ByVal code As String) As Byte() Implements IPrintCommand.Code39
            Return New Byte() {27, "b"c.ToByte, 6}.AddBytes((New Byte() {2})).AddBytes((New Byte() {50})).AddBytes((New Byte() {0})).AddTextBytes(code, Encoding).AddBytes(New Byte() {0}) ' Code Type
            ' Witdh
            ' Height
            ' If print code informed (1 print, 0 dont print)
        End Function

        Public Function Ean13(ByVal code As String) As Byte() Implements IPrintCommand.Ean13
            If code.Trim().Length <> 13 Then Return New Byte(-1) {}
            Return New Byte() {27, "b"c.ToByte(), 1}.AddBytes((New Byte() {2})).AddBytes((New Byte() {50})).AddBytes((New Byte() {0})).AddTextBytes(code.Substring(0, 12), Encoding).AddBytes(New Byte() {0}) ' Code Type
            ' Witdh
            ' Height
            ' If print code informed (1 print, 0 dont print)
        End Function

        Public Function Center() As Byte() Implements IPrintCommand.Center
            Return New Byte() {27, "j"c.ToByte(), 1}
        End Function

        Public Function Left() As Byte() Implements IPrintCommand.Left
            Return New Byte() {27, "j"c.ToByte(), 0}
        End Function

        Public Function Right() As Byte() Implements IPrintCommand.Right
            Return New Byte() {27, "j"c.ToByte(), 2}
        End Function

#End Region

    End Class

End Namespace

File: InnerLibs/Printer/EscDaruma.cs
using InnerLibs.Printer;
using InnerLibs.Printer.Command;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Text;

namespace InnerLibs.EscDarumaCommands
{
    public class EscDaruma : IPrintCommand
    {
        #region Properties

        public int ColsCondensed
        {
            get
            {
                return 57;
            }
        }

        public int ColsExpanded
        {
            get
            {
                return 25;
            }
        }

        public int ColsNormal
        {
            get
            {
                return 48;
            }
        }

        // Encoding.GetEncoding("IBM860")
        public Encoding Encoding { get; set; }

        #endregion Properties

        #region Methods

        private static IEnumerable<byte> StoreQr(string qrData, QrCodeSize size)
        {
            int length = qrData.Length + 3;
            byte b = (byte)(length % 255);
            byte b2 = (byte)Math.Round(length / 255d);
            return (new byte[] { 27, 106, 49 }).AddBytes(new byte[] { 27, 129 }).AddBytes(new[] { b }).AddBytes(new[] { b2 }).AddBytes(new[] { ((int)size + 3).ToByte() }).AddBytes(new[] { 'M'.ToByte() });
        }

        public byte[] AutoTest()
        {
            return new byte[] { 28, 'M'.ToByte(), 254, 0 };
        }

        public byte[] Bold(bool state)
        {
            return state == true ? (new byte[] { 27, 'E'.ToByte() }) : (new byte[] { 27, 'F'.ToByte() });
        }

        public byte[] Center()
        {
            return new byte[] { 27, 'j'.ToByte(), 1 };
        }

        public byte[] Code128(string code)
        {
            return (new byte[] { 27, 'b'.ToByte(), 5 }).AddBytes(new byte[] { 2 }).AddBytes(new byte[] { 50 }).AddBytes(new byte[] { 0 }).AddTextBytes(code, Encoding).AddBytes(new byte[] { 0 }); // Code Type
            // Witdh Height If print code informed (1 print, 0 dont print)
        }

        public byte[] Code39(string code)
        {
            return (new byte[] { 27, 'b'.ToByte(), 6 }).AddBytes(new byte[] { 2 }).AddBytes(new byte[] { 50 }).AddBytes(new byte[] { 0 }).AddTextBytes(code, Encoding).AddBytes(new byte[] { 0 }); // Code Type
            // Witdh Height If print code informed (1 print, 0 dont print)
        }

        public byte[] Condensed(bool state)
        {
            return state == true ? (new byte[] { 27, 15 }) : (new byte[] { 27, 18, 20 });
        }

        public byte[] Ean13(string code)
        {
            if (code.Trim().Length != 13)
                return new byte[0];
            return (new byte[] { 27, 'b'.ToByte(), 1 }).AddBytes(new byte[] { 2 }).AddBytes(new byte[] { 50 }).AddBytes(new byte[] { 0 }).AddTextBytes(code.Substring(0, 12), Encoding).AddBytes(new byte[] { 0 }); // Code Type
            // Witdh Height If print code informed (1 print, 0 dont print)
        }

        public byte[] Expanded(bool state)
        {
            return state == true ? (new byte[] { 27, 'w'.ToByte(), 1 }) : (new byte[] { 27, 'w'.ToByte(), 0 });
        }

        public byte[] FullCut()
        {
            return new byte[] { 27, 'm'.ToByte() };
        }

        public byte[] Initialize()
        {
            return new byte[] { 27, '@'.ToByte() };
        }

        public byte[] Italic(bool state)
        {
            return state == true ? (new byte[] { 27, '4'.ToByte(), 1 }) : (new byte[] { 27, '4'.ToByte(), 0 });
        }

        public byte[] LargeFont()
        {
            return new byte[] { 27, 14, 0 };
        }

        public byte[] LargerFont()
        {
            return new byte[] { 27, 14, 0 };
        }

        public byte[] Left()
        {
            return new byte[] { 27, 'j'.ToByte(), 0 };
        }

        public byte[] NormalFont()
        {
            return new byte[] { 20 };
        }

        public byte[] OpenDrawer()
        {
            return new byte[] { 27, 'p'.ToByte() };
        }

        public byte[] PartialCut()
        {
            return new byte[] { 27, 'm'.ToByte() };
        }

        public byte[] PrintImage(Image image, bool highDensity)
        {
            var list = new List<byte>();
            var bmp = new Bitmap(image);
            string send = InnerLibs.Text.Empty + '\u001b' + '3' + '\0';
            var data = new byte[send.Length];
            for (int i = 0, loopTo = send.Length - 1; i <= loopTo; i++)
                data[i] = Text.ToAscByte(send[i]);
            list.AddRange(data);
            data[0] = Text.ToAscByte('\0');
            data[1] = Text.ToAscByte('\0');
            data[2] = Text.ToAscByte('\0');
            var escBmp = new[] { (byte)0x1B, (byte)0x2A, (byte)0x0, (byte)0x0, (byte)0x0 };
            escBmp[2] = Text.ToAscByte('!');
            escBmp[3] = (byte)(bmp.Width % 256);
            escBmp[4] = (byte)Math.Round(bmp.Width / 256d);
            for (double i = 0d, loopTo1 = bmp.Height / 24d + 1d - 1d; i <= loopTo1; i++)
            {
                list.AddRange(escBmp);
                for (int j = 0, loopTo2 = bmp.Width - 1; j <= loopTo2; j++)
                {
                    for (int k = 0; k <= 24 - 1; k++)
                    {
                        if (i * 24d + k < bmp.Height)
                        {
                            var pixelColor = bmp.GetPixel(j, (int)Math.Round(i * 24d + k));
                            if (!(pixelColor.R > 160 && pixelColor.G > 160 && pixelColor.B > 160))
                                data[(int)Math.Round(k / 8d)] += (byte)(128 >> k % 8);
                            if (highDensity)
                                continue;
                            if (pixelColor.R == 0)
                                data[(int)Math.Round(k / 8d)] += (byte)(128 >> k % 8);
                        }
                    }

                    list.AddRange(data);
                    data[0] = Text.ToAscByte('\0');
                    data[1] = Text.ToAscByte('\0');
                    data[2] = Text.ToAscByte('\0');
                }
            }

            list.AddRange(new byte[] { 27, '@'.ToByte() });
            return list.ToArray();
        }

        public byte[] Printqrdata(string qrData) => ((IPrintCommand)this).PrintQrData(qrData);

        public byte[] PrintQrData(string qrData, QrCodeSize qrCodeSize)
        {
            var list = new List<byte>();
            list.AddRange(StoreQr(qrData, qrCodeSize));
            list.AddRange(Encoding.GetBytes(qrData));
            return list.ToArray();
        }

        public byte[] Right()
        {
            return new byte[] { 27, 'j'.ToByte(), 2 };
        }

        public byte[] Underline(bool state)
        {
            return state == true ? (new byte[] { 27, '-'.ToByte(), 1 }) : (new byte[] { 27, '-'.ToByte(), 0 });
        }

        byte[] IPrintCommand.PrintQrData(string qrData)
        {
            return PrintQrData(qrData, QrCodeSize.Size0);
        }

        #endregion Methods
    }
}